iree/compiler/Conversion/LinalgToSPIRV/CooperativeMatrixAnalysis.cpp

```diff
@@ -0,0 +1,102 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "iree/compiler/Conversion/LinalgToSPIRV/CooperativeMatrixAnalysis.h"
+
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/SPIRV/TargetAndABI.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+
+using namespace mlir;
+
+namespace {
+bool isLegalVectorContract(vector::ContractionOp contract) {
+  if (llvm::size(contract.masks()) != 0) return false;
+  VectorType lhsType = contract.lhs().getType().cast<VectorType>();
+  VectorType rhsType = contract.rhs().getType().cast<VectorType>();
+  VectorType accType = contract.acc().getType().cast<VectorType>();
+
+  std::tuple<int, int, int> dim(lhsType.getDimSize(0), rhsType.getDimSize(1),
+                                lhsType.getDimSize(1));
+  bool supportedType = false;
+  // Check if the mattrix type can be supported as a cooperative matrix.
+  // Currently we have hardcoded checks for what Turing hardware supports.
+  // TODO(thomasraoux): Add device information to be able to query what the
+  // device supports.
+  if (lhsType.getElementType().isInteger(8) &&
+      rhsType.getElementType().isInteger(8) &&
+      accType.getElementType().isInteger(32)) {
+    if (dim == std::make_tuple(8, 8, 32) ||
+        dim == std::make_tuple(16, 16, 32) || dim == std::make_tuple(16, 8, 32))
+      supportedType = true;
+  }
+
+  if (lhsType.getElementType().isF16() && rhsType.getElementType().isF16() &&
+      (accType.getElementType().isF16() || accType.getElementType().isF32())) {
+    if (dim == std::make_tuple(8, 8, 16) ||
+        dim == std::make_tuple(16, 16, 16) || dim == std::make_tuple(16, 8, 16))
+      supportedType = true;
+  }
+  if (!supportedType) return false;
+  return true;
+}
+
+bool supportsCooperativeMatrix(Operation* op) {
+  if (isa<vector::TransferReadOp>(op) || isa<vector::TransferWriteOp>(op))
+    return true;
+  if (isa<vector::ContractionOp>(op) &&
+      isLegalVectorContract(cast<vector::ContractionOp>(op)))
+    return true;
+  // We only support minimal set of operations right now. We cna trivially
+  // extend to ALU instructions supporint Cooperative Matrix in SPIR-V spec.
+  // We also need to extend to control flow operations, Alloca, etc...
+  // TODO(thomasraoux): extend support to more complex chain of instructions.
+  return false;
+}
+}  // namespace
+namespace mlir {
+namespace iree_compiler {
+
+CooperativeMatrixAnalysis::CooperativeMatrixAnalysis(mlir::Operation* op) {
+  auto targetEnv = spirv::TargetEnv(spirv::lookupTargetEnv(op));
+  if (!targetEnv.allows(spirv::Capability::CooperativeMatrixNV) ||
+      !targetEnv.allows(spirv::Extension::SPV_NV_cooperative_matrix))
+    return;
+
+  op->walk([&](Operation* op) {
+    auto contract = dyn_cast<vector::ContractionOp>(op);
+    if (contract == nullptr) return;
+    auto hasNoVectorDest = [](Operation* op) {
```

Comment:
It would be nice to name the `op` differently here. Shadowing variables increases the burden on understanding the logic and sometimes it can leads to oversight and thus bugs. :)